op-conductor/health/monitor_test.go

```diff
@@ -26,130 +26,192 @@ const (
-		return &eth.SyncStatus{
-			UnsafeL2: eth.L2BlockRef{
-				Time: unsafeTime,
-			},
-			SafeL2: eth.L2BlockRef{
-				Time: now,
-			},
-		}
-	}
-	s.rc.ExpectSyncStatus(syncStatusGenerator(now), nil)
-	s.rc.ExpectSyncStatus(syncStatusGenerator(now), nil)
-	s.rc.ExpectSyncStatus(syncStatusGenerator(now+2), nil)
-	s.rc.ExpectSyncStatus(syncStatusGenerator(now+2), nil)
-	s.rc.ExpectSyncStatus(syncStatusGenerator(now+4), nil)
-	s.rc.ExpectSyncStatus(syncStatusGenerator(now+4), nil)
-
-	healthUpdateCh := s.monitor.Subscribe()
-	for i := 0; i < 6; i++ {
+
+	rc := &testutils.MockRollupClient{}
+	rc.ExpectSyncStatus(mockSyncStatus(now, 1, now, 1), nil)
+	rc.ExpectSyncStatus(mockSyncStatus(now, 1, now, 1), nil)
+	rc.ExpectSyncStatus(mockSyncStatus(now+2, 2, now, 1), nil)
+	rc.ExpectSyncStatus(mockSyncStatus(now+2, 2, now, 1), nil)
+	rc.ExpectSyncStatus(mockSyncStatus(now+4, 3, now, 1), nil)
+	rc.ExpectSyncStatus(mockSyncStatus(now+4, 3, now, 1), nil)
+
+	monitor := s.SetupMonitor(now, 60, 3, rc, nil)
+	healthUpdateCh := monitor.Subscribe()
+
+	for i := 0; i < 5; i++ {
 		healthy := <-healthUpdateCh
-		if i < 5 {
+		if i < 4 {
 			s.Nil(healthy)
 		} else {
 			s.NotNil(healthy)
 		}
 	}
+
+	s.NoError(monitor.Stop())
+}
+
+func (s *HealthMonitorTestSuite) TestHealthyWithUnsafeLag() {
+	s.T().Parallel()
+	now := uint64(time.Now().Unix())
+
+	rc := &testutils.MockRollupClient{}
+	// although unsafe has lag of 20 seconds, it's within the configured unsafe interval
+	// and it is advancing every block time, so it should be considered safe.
+	rc.ExpectSyncStatus(mockSyncStatus(now-10, 1, now, 1), nil)
+	rc.ExpectSyncStatus(mockSyncStatus(now-10, 1, now, 1), nil)
+	rc.ExpectSyncStatus(mockSyncStatus(now-8, 2, now, 1), nil)
+	rc.ExpectSyncStatus(mockSyncStatus(now-8, 2, now, 1), nil)
+
+	monitor := s.SetupMonitor(now, 60, 60, rc, nil)
+	healthUpdateCh := monitor.Subscribe()
+
+	// confirm initial state
+	s.Zero(monitor.lastSeenUnsafeNum)
+	s.Zero(monitor.lastSeenUnsafeTime)
+
+	// confirm state after first check
+	healthy := <-healthUpdateCh
+	s.Nil(healthy)
+	lastSeenUnsafeTime := monitor.lastSeenUnsafeTime
+	s.NotZero(monitor.lastSeenUnsafeTime)
+	s.Equal(uint64(1), monitor.lastSeenUnsafeNum)
+
+	healthy = <-healthUpdateCh
+	s.Nil(healthy)
+	s.Equal(lastSeenUnsafeTime, monitor.lastSeenUnsafeTime)
+	s.Equal(uint64(1), monitor.lastSeenUnsafeNum)
+
+	healthy = <-healthUpdateCh
+	s.Nil(healthy)
+	s.Equal(lastSeenUnsafeTime+2, monitor.lastSeenUnsafeTime)
+	s.Equal(uint64(2), monitor.lastSeenUnsafeNum)
+
+	s.NoError(monitor.Stop())
+}
+
+func mockSyncStatus(unsafeTime, unsafeNum, safeTime, safeNum uint64) *eth.SyncStatus {
+	return &eth.SyncStatus{
+		UnsafeL2: eth.L2BlockRef{
+			Time:   unsafeTime,
+			Number: unsafeNum,
+		},
+		SafeL2: eth.L2BlockRef{
+			Time:   safeTime,
+			Number: safeNum,
+		},
+	}
 }
 
 func TestHealthMonitor(t *testing.T) {
 	suite.Run(t, new(HealthMonitorTestSuite))
 }
+
+type timeProvider struct {
```

Comment:
[nit] I believe there's already a fake clock utility in `op-service` you can use

Reply:
Thanks, just checked `DeterministicClock`, looks good. Can rework in a later PR